Introduce commit subset check (part of stack)

Caleb-T-Owens opened the pull request:
Introduces a function for determining if the changes introduced by a commit, are a subset of the changes introduced by another.

This is helpful for an upcoming PR which should greatly improve our integration check.

---

`get_exclusive_tree` is able to take a stack of commits IE `Base -> A -> B`. And return a tree which only contains the introductions of the `B` commit (when diffed against the Base).

We do this by merging commits `B` and `Base` against the parent of `B`. Any conflicts are resolved in favour of the `B` commit.

`is_subset` makes use of `get_exclusive_tree` to be able to compare the subset and superset candidates without worry about changes that they may have been cherry-picked on top of

<!-- GitButler Footer Boundary Top -->
---
This is **part 1 of 2 in a stack** made with GitButler:
- <kbd>&nbsp;2&nbsp;</kbd> #5893 
- <kbd>&nbsp;1&nbsp;</kbd> #5881 👈 
<!-- GitButler Footer Boundary Bottom -->

Caleb-T-Owens committed:
Introduce commit subset check

Byron committed:
minimal refactor

This is more of a suggestion than anything else.

Byron committed:
When checking for conflicts, always be explicit about what's a conflict.

Byron commented on crates/gitbutler-branch-actions/src/commit_ops.rs:
```diff
@@ -0,0 +1,238 @@
+        use super::*;
+
+        #[test]
+        fn a_commit_is_a_subset_of_itself() {
```
All these tests are read-only, which is when the `read_only::fixture*` methods are faster. They create a fixture only once and make it easy to inspect on disk. They come at the disadvantage of separating the test-setup from the test itself though, but to me it's easy enough to check the script and I can understand more easily how the structure looks like by looking at it, and if in doubt the final structure is easy to inspect on disk.

Of course, these are not the pinnacle of what's possible and I always thought that it would be great to define tests more like `cargo` does, but to apply the same techniques to easily define complex Git repositories.

Byron reviewed (commented):
Thanks for the invite, I took a close look, but only from a Rust + `gix` perspective, not from a "does this really work and is it sound" point of view.

For simplicity, I just created two commits with changes I'd do, please see them as suggestions. The second one has more of a "careful, better do it this way" narrative though, but either way they can be changed, refactored or removed.

Generally, when doing checks like these, one should assure that the `Repository` is configured to *not* write objects to disk. This is fast enough to do it on the fly, but as it would require cloning the repository that also strips object caches which would hurt performance if many commits are checked in a row. Thus I added documentation to the method that I think might be public one day, but ideally the top-level caller will take care of that.

I hope this helps.

--- 

As a final something to think about: I personally find it helpful to know about lifetimes despite the added noise, and enable `#![deny(rust_2018_idioms)]` in `lib.rs` in all crates. This came up when seeing `get_first_parent(commit: gix::Commit)` and I had to wonder why it's passed like this, until I realized that it does have a reference to the `Repository` which was just hidden. With `get_first_parent(commit: gix::Commit<'_>)` it would be clear that contains references, just like the `git2` types of the same kind. When passing it as `&` which should be done with a `Clone` type, it also became obvious why it was done like that - having all lifetimes visible may have helped here, too.

Caleb-T-Owens commented on crates/gitbutler-oxidize/src/lib.rs:
```diff
@@ -101,3 +121,16 @@ pub fn gix_to_git2_index(index: &gix::index::State) -> anyhow::Result<git2::Inde
+
+pub fn print_tree(tree: gix::Tree<'_>) {
+    let mut recorder = gix::traverse::tree::Recorder::default();
+    tree.traverse().breadthfirst(&mut recorder).unwrap();
```
I wasn't too concerned with the ordering/bredthfirstness of the function. In some ways, it's possibly more helpful if it prints in same order as the tree entries.

I'll have a think about what a helpful debug output would look like and revisit

Byron replied:
Within `#[test]` code `dev-dependencies` are available, no matter where it is.
Thus, moving it to the `testsupport` crate will allow it to be used in any test, anywhere, as long as `testsupport` is a `dev-dependency`.

Caleb-T-Owens commented on crates/gitbutler-oxidize/src/lib.rs:
```diff
@@ -101,3 +121,16 @@ pub fn gix_to_git2_index(index: &gix::index::State) -> anyhow::Result<git2::Inde
     Ok(out)
 }
+
+pub fn print_tree(tree: gix::Tree<'_>) {
```
I was wanting to use it for debugging inside an application function rather than inside a test case, which afaik, means it can't be inside `testsupport`

Byron replied:
Within `#[test]` code `dev-dependencies` are available, no matter where it is.
Thus, moving it to the `testsupport` crate will allow it to be used in any test, anywhere, as long as `testsupport` is a `dev-dependency`.

Caleb-T-Owens commented:
> Thanks for the invite, I took a close look, but only from a Rust + gix perspective, not from a "does this really work and is it sound" point of view.

Thank you! This is greatly appreciated

> For simplicity, I just created two commits with changes I'd do, please see them as suggestions. The second one has more of a "careful, better do it this way" narrative though, but either way they can be changed, refactored or removed.

Thanks for these! Both look good.

> Generally, when doing checks like these, one should assure that the Repository is configured to not write objects to disk. This is fast enough to do it on the fly, but as it would require cloning the repository that also strips object caches which would hurt performance if many commits are checked in a row. Thus I added documentation to the method that I think might be public one day, but ideally the top-level caller will take care of that.

The new integration check will run over a full stack rather than getting called on an individual commit. As such, it will almost always be called many times in a row. As such, it makes more sense to open the in-memory repository outside of the `is_subset` function and instead inside the new integration function.

> As a final something to think about: I personally find it helpful to know about lifetimes despite the added noise, and enable #![deny(rust_2018_idioms)] in lib.rs in all crates. This came up when seeing get_first_parent(commit: gix::Commit) and I had to wonder why it's passed like this, until I realized that it does have a reference to the Repository which was just hidden. With get_first_parent(commit: gix::Commit<'_>) it would be clear that contains references, just like the git2 types of the same kind. When passing it as & which should be done with a Clone type, it also became obvious why it was done like that - having all lifetimes visible may have helped here, too.

I'm wary of becoming a perl codebase, but would it make sense to add `#![deny(rust_2018_idioms)]` to the top of new files so we can incrementally move over to better idioms?

Byron commented:
> I'm wary of becoming a perl codebase, but would it make sense to add `#![deny(rust_2018_idioms)]` to the top of new files so we can incrementally move over to better idioms?

It absolutely can be done that way, which certainly helps to get a feel for it. Additionally, I still don't understand how crates specified as `edition = "2021"` won't even warn about the rust-2018 idioms.
The reason it's not the default might be exactly due to the added noise, even though I find it most useful to know if references are involved.

Caleb-T-Owens committed:
Use an enum result to better describe subset
